concent_api/core/models.py

```diff
@@ -73,3 +75,216 @@ class MessageAuth(Model):
     message              = OneToOneField(StoredMessage, related_name = 'auth')
     provider_public_key  = Base64Field(max_length=64)
     requestor_public_key = Base64Field(max_length=64)
+
+
+class ClientManager(Manager):
+
+    def get_or_create_full_check(self, public_key):
+        """
+        Returns Model instance.
+        Does the same as get_or_create method, but also performs full_clean() on newly created instance.
+        """
+        try:
+            instance = self.get(public_key = public_key)
+        except self.model.DoesNotExist:
+            instance = self.model(
+                public_key = public_key
+            )
+            instance.full_clean()
+            instance.save()
+        return instance
+
+
+class Client(models.Model):
+    """
+    Represents Concent client, identified by public key.
+    """
+
+    objects = ClientManager()
+
+    public_key = Base64Field(max_length = 64, unique = True)
+
+
+class Subtask(models.Model):
+    """
+    Represents subtasks states.
+    """
+
+    class SubtaskState(ChoiceEnum):
+        ForcingReport           = 'forcing_report'
+        Reported                = 'reported'
+        ForcingResultTransfer   = 'forcing_result_transfer'
+        ResultUploaded          = 'result_uploaded'
+        ForcingAcceptance       = 'forcing_acceptance'
+        Rejected                = 'rejected'
+        AdditionalVerification  = 'additional_verification'
+        Accepted                = 'accepted'
+        Failed                  = 'failed'
+
+    class Meta:
+        unique_together = (
+            ('requestor', 'task_id'),
+            ('requestor', 'subtask_id'),
+        )
+
+    task_id                     = models.CharField(max_length = MESSAGE_TASK_ID_MAX_LENGTH)
+    subtask_id                  = models.CharField(max_length = MESSAGE_TASK_ID_MAX_LENGTH)
+
+    # Relation to concent client in requestor context.
+    requestor                   = models.ForeignKey(Client, related_name = 'subtasks_as_requestor')
+
+    # Relation to concent client in provider context.
+    provider                    = models.ForeignKey(Client, related_name = 'subtasks_as_provider')
+
+    state                       = models.CharField(max_length = 32, choices = SubtaskState.choices())
```

Comment:
You should have an assert that no value from the enum is longer than 32 characters.

Also, what I usually do with enums is to use single-letter values so that we can just have a column of length 1. They have names anyway (accessible via the `name` property) so values are really redundant.

Reply:
:+1: I will add validation, but I am against single-letters, they are not descriptive, and gets even worse if you have several starting with the same letter. You end up having query like:
`.filter(column = enum.Z) and you don't actually know what 'Z' is without checking.

Reply:
Yeah, I often used ints for this in the past but later I switched to single-letter constants because:
- With ints you usually use something like [`AutoNumber`](https://docs.python.org/3/library/enum.html#using-a-custom-new). This makes it easy for someone to come later and add a new value somewhere in the middle without realizing that this shifts all the values after it and requires a database migration.
- In case you need to work with the raw content of the field (e.g. in fixtures) the single-letter values are a bit easier to decipher than just ints and it's harder to make a mistake.
- They're just as small as ints in terms of storage.

I'm not saying it's a perfect solution but I personally like it a bit more than storing whole string values. It would be best if we could use have enums in the database. [PostgreSQL supports enums natively](https://www.postgresql.org/docs/9.1/static/datatype-enum.html) but [Django can't use them (yet)](https://code.djangoproject.com/ticket/24342).

Reply:
I have skipped it on purpose because I wanted to get back to this. 

> At least please add a method in ChoiceEnum that computes the maximum length of all values of a particular enum and use its result here instead of hard-coding 32.

How would this work and when should this function run ? If someone will add entry to enum extending the current length, the code will update the value at runtime, but it won't change because it needs creating a migration with that value not in runtime.